content/articles/how-to-build-a-web-crawler-with-java/index.md

```diff
@@ -0,0 +1,178 @@
+Feedback monitoring: You can also use web crawlers to traverse the web in search of where users may have mentioned the name of your company for reviews. 
+
+### Challenges of building a Web crawler.
+
+As much as web crawlers come with many benefits, they tend to pose some challenges when building them. Some of the issues faced include:
+
+Server overload: This commonly occurs when the crawler traverses irrelevant web pages. When the crawler navigates a vast number of web pages, it impacts the server's performance.
+Presence of Anti-Scraping tools: Organizations are beginning to integrate anti-scraping tools into their website to prevent unauthorized visits to their web pages. These tools can distinguish bots from humans and restrict access to the bots from carrying out malicious activities on their web pages.
+Inconsistent webpage structures: The structures of websites are different from one another. Due to this dynamism, a web crawler that performs well on one website may fail on another. 
+
+### Java libraries for building web crawlers
+
+Although this tutorial will only cover the concept of web crawling at the fundamental level without the use of any external libraries, here are some Java API’s you can integrate with your to perform web crawling:
+
+* [Heritrix](https://github.com/internetarchive/heritrix3)
+* [JSoup](https://jsoup.org/)
+* [Apache nutch](https://nutch.apache.org/)
+* [Stormcrawler](http://stormcrawler.net/)
+* [Gecco](https://github.com/xtuhcy/gecco)
+
+### Breadth-First Search - The underlying algorithm of the Web crawler
+
+Before proceeding further with this tutorial, you must have a fundamental knowledge of the breadth-first search (BFS) algorithm that we’ll implement in building the web crawler in the subsequent sections. It’s okay to skip this section if you are familiar with how BFS operates.
+
+BFS is a traversal algorithm you implement when you want to visit every node of a tree or graph data structure in a layerwise format. This implies that given a source node, BFS visits all the children of the source node horizontal (either left to right or right to left, depending on how you choose to implement it), then it goes on to visit the next layer of children, that is, the “grandchildren” of the source node. 
+
+![breadth-first search](/bfs.png)
+
+In the diagram above, the traversal begins from Node 1, the parent or source node. Then it proceeds to its children - Nodes 2, 3, 4 in the second layer. From Node 4, the traversal continues to the leftmost part of the third layer and travels horizontally to the end of the layer. This continues until all nodes are visited.
+
+BFS has a time complexity of O(V+E), where V denotes the number of vertices and E, the number of edges. 
```

Comment:
Since you've touched upon the complexity part - you'll have to explain these too - at least a short summary of what it is.

And, add reference links from Section website.